WT-283 Add WT_SESSION::alter method

sueloverso opened the pull request:
@agorrod Here's a branch for the alter command.  I added a test to alter and modify all combinations of both `access_pattern_hint` and `cache_resident`.  Can you think of other testing we should add?  You should take a first pass at reviewing this.

agorrod commented:
autobuild: retest this please

daveh86 commented:
retest this please

daveh86 commented:
retest this please

sueloverso commented:
This is not yet ready to merge as I need to test/debug LSM, column groups, indexes, etc.

agorrod commented:
Thanks @sueloverso this is looking good to me so far.

> Can you think of other testing we should add?

* Testing different types of tables, including those with complex schema as you pointed out.
* Testing that checkpoint, backup, verify and compact can run in parallel without any hangs or unexpected problems.
* Testing that the expected value is persisted to disk. i.e: it is still the expected setting if re-opening a table after the connection has been reopened.
* Testing backward/forward compatibility. I'm not sure what the expectations are, but I'd use testing to flush it out.

One other piece I'd like to see in this branch is exposing alter via the command line utility.

sueloverso commented:
@agorrod and @michaelcahill I believe this is ready for review and merge.  I added it to test/format and debugged some failures there.  In particular, please consider the LSM changes - I had to make the call to `wt_lsm_tree_get` non-exclusive.  I added data-source support today too.

sueloverso commented:
FYI, I ran test/format overnight, forcing `alter=1` all the time and it was still happy this morning.

ddanderson commented on src/include/wiredtiger.in:
```diff
@@ -800,6 +800,34 @@ struct __wt_session {
+	 * @snippet ex_all.c Alter a table
+	 *
+	 * @param session the session handle
+	 * @param name the URI of the object to drop, such as \c "table:stock"
```
@sueloverso, this should be _object to alter_, not _object to drop_.

sueloverso replied:
Fixed! Its provenance was showing. :-)

michaelcahill commented on src/schema/schema_alter.c:
```diff
@@ -0,0 +1,190 @@
+	 */
+	tblcfg[0] = table->config;
+	tblcfg[1] = NULL;
+	WT_ERR(__wt_table_colgroups_config(session, tblcfg, &ncolgroups));
```
I believe this code can just check `table->ncolgroups == 0` because that is resolved by `__wt_schema_get_table`.  Is there some more complicated case I'm missing?

michaelcahill commented on test/format/config.h:
```diff
@@ -65,6 +65,10 @@ static CONFIG c[] = {
+	{ "alter",
+	  "if altering the table is enabled",			/* 50% */
+	  C_BOOL, 50, 0, 0, &g.c_alter, NULL },
+
```
Maybe only turn alter on in 5% of `test/format` runs?

michaelcahill commented on src/docs/command-line.dox:
```diff
@@ -37,6 +37,39 @@ In general, commands that modify the database or tables will run recovery
+
+@code
+table cache_resident=true
+table:xxx cache_resident=true
```
Can the documentation please be in terms of `access_pattern_hint`?  As discussed, I think we may want to deprecate `cache_resident` at some point...
